python/pyspark/sql/dataframe.py

```diff
@@ -3928,6 +3928,71 @@ def dropDuplicates(self, subset: Optional[List[str]] = None) -> "DataFrame":
             jdf = self._jdf.dropDuplicates(self._jseq(subset))
         return DataFrame(jdf, self.sparkSession)
 
+    def dropDuplicatesWithinWatermark(self, subset: Optional[List[str]] = None) -> "DataFrame":
```

Comment:
Can we have one `*cols`? I believe that's more consistent with other API in this file. e.g., groupby, select, etc.

Reply:
This is somewhat similar with https://github.com/apache/spark/pull/40561#discussion_r1159347699 too. Having one `*cols` would allow both `df.dropDuplicatesWithinWatermark(col1, col2)` and `dropDuplicatesWithinWatermark(*cols)`

Reply:
Grouping to the above Scala API in question; it is important to decide which one we want this API to be consistent with. I'm seeing this API as a sibling of existing dropDuplicates API, which I wish to provide the same UX.

Say, for the case of batch query, someone can add the postfix of the name from existing query (dropDuplicates -> dropDuplicatesWithinWatermark) and expect the query to be run without any issue. It is respected now regardless of the type of parameter they use. 

For the case of streaming query, they might already have a query dealing with dropDuplicates but had been suffered from the issue I've mentioned in the section `why are the changes needed?`. In most cases, we expect them to just add the postfix of the name from existing query and discard checkpoint, and then enjoy that the problem gets fixed. I hope this is enough rationale to make the new API be consistent with dropDuplicates.

Reply:
Oh, I see. Okay, makes sense to me.

Reply:
Thanks!